src/Primitives/src/StringValues.cs

```diff
@@ -66,51 +85,92 @@ public StringValues(string[] values)
 
         public string this[int index]
         {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
             get
             {
-                if (_values != null)
+                // Take local copy of _values so type checks remain valid even if the StringValues is overwritten in memory
+                var value = _values;
+                if (index == 0 && value is string str)
                 {
-                    return _values[index]; // may throw
+                    return str;
                 }
-                if (index == 0 && _value != null)
+                else if (value != null)
```

Comment:
Is it better to move the  `value is null` case here?

Reply:
Pushed it out of the method as hopefully it shouldn't ever happen (as you are accessing out of bounds); and reduces the agressive inline size